### PR TITLE
Update analyzer mentioned files

### PR DESCRIPTION
**Pull request by martong**
* Since https://github.com/Ericsson/clang/pull/199/files we can track which source files corresponding AST had been loaded before a crash.
Therefore, it is no longer needed to have complex regex matching in `get_analyzer_mentioned_files`.
* We put absolute paths into the failed zip under sources-root, because all analyzer runs are using absolute paths.

### CODE REVIEW

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py [hunk 1]
```diff
@@ -160,35 +157,35 @@ def construct_analyzer_cmd(self, result_handler):
...
-                line = line.replace("In file included from ", "", 1)
+        xtu_ast_dir = os.path.join(self.get_xtu_dir(), "ast")
 
-            match = re.match(regex, line)
+        for line in output.splitlines():
+            match = re.match(regex_for_ctu_ast_load, line)
             if match:
-                paths.append(match.group('path'))
+                path = match.group(1)
+                assert path.find(xtu_ast_dir) != -1
```
Comment (whisperity) [inline libcodechecker/analyze/analyzers/analyzer_clangsa.py]: Certain python environments can overoptimise code which run with assertions disabled. Also, a Python assert doesn't really carry information. An Exception, and a proper, user-understandable (or at least, via the exception, developer-understandable) error message should be used.

Is this assertion needed in the first place?

Why do we use `find` instead of the `substr in string` syntax?

File: tests/functional/ctu_failure/__init__.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+# Test workspace should be initialized in this module.
+TEST_WORKSPACE = None
+
+
+def setup_package():
+    """Setup the environment for testing ctu."""
+
+    global TEST_WORKSPACE
+    TEST_WORKSPACE = env.get_workspace('ctu_failure')
```
Comment (whisperity) [inline tests/functional/ctu_failure/__init__.py]: We have a **lot** of subdirectories under `functional` already. Can't the test module file of this test package instead go as a new python module in the `ctu` folder?

File: tests/functional/ctu_failure/test_ctu_failure.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        """ Test the failure zip contains the source of imported TU"
+        """
+
+        if not self.ctu_capable:
+            self.skipTest(NO_CTU_MESSAGE)
+
+        output = self.__do_ctu_all(reparse=False,
+                                   extra_args=[
+                                       "--verbose", "debug",
+                                       "-e", "debug.ExprInspection"
```
Comment (gyorb) [inline tests/functional/ctu_failure/test_ctu_failure.py]: Please add a comment why do we enable/use this checker for this test.

File: tests/functional/ctu_failure/test_files/lib.c [hunk 4]
```diff
@@ -0,0 +1,5 @@
+int foo(int i)
```
Comment (gyorb) [inline tests/functional/ctu_failure/test_files/lib.c]: We have a separate directory for the [test_projects](https://github.com/Ericsson/codechecker/tree/09837984d59daecb1f799290a06eb9a105c6020b/tests/projects) which will be analyzed during the tests.

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py [hunk 5]
```diff
@@ -160,35 +157,35 @@ def construct_analyzer_cmd(self, result_handler):
...
+                                         config.ld_lib_path_extra)
+        triple_arch = ctu_triple_arch.get_triple_arch(self.buildaction,
+                                                      self.source_file,
+                                                      config, env)
+        xtu_dir = os.path.join(config.ctu_dir, triple_arch)
+        return xtu_dir
+
     def get_analyzer_mentioned_files(self, output):
         """
         Parse ClangSA's output to generate a list of files that were mentioned
```
[the remark this answers is not in the document]
Comment (whisperity) [reply libcodechecker/analyze/analyzers/analyzer_clangsa.py]: Technically, yes. It still parses the output, just with a different pattern.

[the remark this answers is not in the document]
Comment (Xazax-hun) [reply libcodechecker/analyze/analyzers/analyzer_clangsa.py]: But it does not collect every mentioned file in the warnings but only files mentioned by a specific log message.

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py [hunk 6]
```diff
@@ -128,15 +128,12 @@ def construct_analyzer_cmd(self, result_handler):
...
-                env = analyzer_env.get_check_env(config.path_env_extra,
-                                                 config.ld_lib_path_extra)
-                triple_arch = ctu_triple_arch.get_triple_arch(self.buildaction,
-                                                              self.source_file,
-                                                              config, env)
                 analyzer_cmd.extend(['-Xclang', '-analyzer-config',
                                      '-Xclang',
-                                     'xtu-dir=' + os.path.join(config.ctu_dir,
-                                                               triple_arch)])
+                                     'xtu-dir=' + self.get_xtu_dir(),
```
Comment (gerazo) [inline libcodechecker/analyze/analyzers/analyzer_clangsa.py]: We've inherited "XTU" name from somewhere else so we made the decision to call everything we do "CTU". Although the legacy parameter here (which was not changed in clang yet) says XTU, we should rather consequently call these CTU in the code. Someone may change it in clang as well in the future. :)

File: tests/functional/ctu_failure/__init__.py (hunk 2 above)
Comment (martong) [reply tests/functional/ctu_failure/__init__.py]: If I understand the structure of the test dirs correctly then each dir represent a test suite with its own setup and optionally files for the test project. And I think, the ctu_failure directory perfectly fits into this structure.
```
   ▾ delete_runs/            
       __init__.py           
       test_delete_runs.py   
   ▾ detection_status/       
       __init__.py           
       test_detection_status.
   ▾ diff/                   
       __init__.py           
       test_diff.py          
   ▾ func_template/          
       template__init__.py   
       template_test.py      
   ▾ instance_manager/       
       __init__.py           
       test_instances.py     
```

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py (hunk 1 above)
Comment (martong) [reply libcodechecker/analyze/analyzers/analyzer_clangsa.py]: OK , I removed the assert and replaced with a condition.

File: tests/functional/ctu_failure/__init__.py (hunk 2 above)
Comment (whisperity) [reply tests/functional/ctu_failure/__init__.py]: I don't think there ever was any structural decision on the functional test folder. The [suppress tests before review status was implemented](https://github.com/Ericsson/codechecker/tree/v5.8/tests/functional/suppress), for example, also tested multiple features at once.

The current layout of the test suite better represents testing individual features or feature sets of CodeChecker. Analyzer handling, analyzer result handling, result querying (report viewer api), diffing, run deletion, product management, etc.

"CTU failure" is not a feature set, it is a new, special test case for CTU. Of course, due to this test being related to analyzer handling, it *could* go into the `analyze` folder, but `CTU` is a bigger, and thus more powerful category.

Comment (whisperity) [reply tests/functional/ctu_failure/__init__.py]: Structurally, nosetest test on the module level, the package level things only handle analysis execution, result storage and server setup.

For example, the current layout is also bad, as every single damn test package runs a full analysis, sometimes on a big project, but at least where it wasn't neccessarry, we cut down on using a simpler project only, thus further quickening up the tests without losing functionality.

Comment (martong) [reply tests/functional/ctu_failure/__init__.py]: Ok, I moved the ctu_failure dir under ctu.

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py [hunk 7]
```diff
@@ -160,35 +157,35 @@ def construct_analyzer_cmd(self, result_handler):
             LOG.error(ex)
             return []
 
+    def get_xtu_dir(self):
+        config = self.config_handler
```
[the remark this answers is not in the document]
Comment (martong) [reply libcodechecker/analyze/analyzers/analyzer_clangsa.py]: Ok, I added a desc.

File: tests/functional/ctu_failure/test_ctu_failure.py (hunk 3 above)
Comment (martong) [reply tests/functional/ctu_failure/test_ctu_failure.py]: Ok , I added a comment.

File: tests/functional/ctu_failure/test_files/lib.c (hunk 4 above)
Comment (martong) [reply tests/functional/ctu_failure/test_files/lib.c]: Ok, I moved it under the projects dir.

File: libcodechecker/analyze/analysis_manager.py [hunk 8]
```diff
@@ -359,6 +359,8 @@ def check(check_data):
 
                     LOG.debug("Writing dependent files to archive.")
                     for dependent_source in dependencies:
+                        if not os.path.isabs(dependent_source):
+                            dependent_source = os.path.abspath(dependent_source)
```
[the remark this answers is not in the document]
Comment (martong) [reply libcodechecker/analyze/analysis_manager.py]: The point here is to replicate the host's file system under the sources-root directory.
`dependent_source` contains the list of files reported as dependent by `$CXX -M`, this includes the arg as well.
```
> gcc -E -M -MT __dummy -c main.c
__dummy: main.c /usr/include/stdc-predef.h

> gcc -E -M -MT __dummy -c $(realpath main.c)                                                  
__dummy: /home/egbomrt/WORK/CodeChecker/tests/projects/ctu_failure/main.c \
 /usr/include/stdc-predef.h
```
We assemble the gcc commandline based on the compile database, but the `directory` key from it is not used, therefore the arg will be a relative path (in the example above it is the first case, just `main.c`). 
The cwd did not changed after calling gcc, so we can call `os.path.abspath()` on the arg thus resulting the correct absolute path.

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (martong) [reply libcodechecker/analyze/analyzers/analyzer_clangsa.py]: We don't need anymore to catch the warnings. If there would be a warning from a TU, that must had been imported by Clang previously (together with the proper `CTU loaded` log).

[the remark this answers is not in the document]
Comment (martong) [reply libcodechecker/analyze/analyzers/analyzer_clangsa.py]: perhaps a better name could be: `get_analyzer_ctu_loaded_files` ?

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py [hunk 9]
```diff
@@ -128,15 +128,12 @@ def construct_analyzer_cmd(self, result_handler):
...
-                triple_arch = ctu_triple_arch.get_triple_arch(self.buildaction,
-                                                              self.source_file,
-                                                              config, env)
                 analyzer_cmd.extend(['-Xclang', '-analyzer-config',
                                      '-Xclang',
-                                     'xtu-dir=' + os.path.join(config.ctu_dir,
-                                                               triple_arch)])
+                                     'xtu-dir=' + self.get_xtu_dir(),
+                                     '-Xclang',
+                                     '-analyzer-display-ctu-progress',
```
[the remark this answers is not in the document]
Comment (martong) [reply libcodechecker/analyze/analyzers/analyzer_clangsa.py]: Good point. The analysis will fail if the used Clang does not support this switch.
This will cause a problem for CodeChecker in the time window when the CTU patch is accepted but the patch for this switch is still not.
I hope this time window will be short, since the patch for `-analyzer-display-ctu-progress` is very tiny. Still this is a valid danger.
We have the following options:
* Create a framework to discover feature capabilities of the used Clang (this is a huge work, counting in men weeks I estimate)
* If the command fails, then retry without `-analyzer-display-ctu-progress` switch
* Any other ideas?

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (martong) [reply libcodechecker/analyze/analyzers/analyzer_clangsa.py]: We can't just rename this function, because this is an inherited abstract function from `SourceAnalyzer`. And it is used in `analyzer_clang_tidy.py` as well.

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py (hunk 6 above)
Comment (gyorb) [reply libcodechecker/analyze/analyzers/analyzer_clangsa.py]: A `FIXME` or a comment would be nice here, that this parameter needs to be changed, or we should create a separate issue.

**Comment by martong** [comment]
@Xazax-hun  @gyorb @whisperity I think I addressed all your comments, could you please review again?

File: libcodechecker/analyze/analyzers/analyzer_clangsa.py (hunk 6 above)
Comment (martong) [reply libcodechecker/analyze/analyzers/analyzer_clangsa.py]: https://github.com/Ericsson/clang/issues/208

File: libcodechecker/analyze/host_check.py [hunk 10]
```diff
@@ -35,3 +36,29 @@ def check_clang(compiler_bin, env):
...
+        proc = subprocess.Popen(cmd,
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE,
+                                )
+        out, err = proc.communicate()
+
+        LOG.debug("stdout:\n" + out)
+        LOG.debug("stderr:\n" + err)
+
+        rc = proc.returncode
```
Comment (Xazax-hun) [inline libcodechecker/analyze/host_check.py]: You could just `return proc.returncode == 0`

File: libcodechecker/analyze/host_check.py [hunk 11]
```diff
@@ -35,3 +36,26 @@ def check_clang(compiler_bin, env):
             LOG.error(oerr)
             LOG.error('Failed to run: "' + ' '.join(clang_version_cmd) + '"')
             return False
+
+
+def has_analyzer_feature(clang_bin, feature):
+    inputFile = tempfile.NamedTemporaryFile()
```
Comment (gyorb) [inline libcodechecker/analyze/host_check.py]: the inputFile should be closed or you can use it in a with statement

File: tests/functional/ctu/ctu_failure/test_ctu_failure.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+                                         cwd=self.test_dir, env=self.env)
+        self.ctu_capable = '--ctu-' in output
+        print("'analyze' reported CTU-compatibility? " + str(self.ctu_capable))
+
+        self.ctu_has_analyzer_display_ctu_progress = \
+            host_check.has_analyzer_feature(self.__getClangSaPath(),
+                                            '-analyzer-display-ctu-progress')
+        print("Has -analyzer-display-ctu-progress? " + str(self.ctu_capable))
+
+        # Fix the "template" build JSONs to contain a proper directory
```
Comment (gyorb) [inline tests/functional/ctu/ctu_failure/test_ctu_failure.py]: I would recommend to copy the whole `ctu_failure` test project under the test workspace ad do the modifications there and not in the source tree. Later some other test might use this project which could cause conflicts.
Maybe less `os.chdir` is needed with the test project copy. I'm not a big fan of it, it can be hard to follow in which directory the scripts are running.

File: tests/functional/host_check/__init__.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+import shutil
+
+from libtest import env
+
+# Test workspace should be initialized in this module.
+TEST_WORKSPACE = None
+
+
+def setup_package():
+    """Setup the environment for testing ctu."""
```
Comment (gyorb) [inline tests/functional/host_check/__init__.py]: This is not a ctu test.

File: tests/functional/host_check/__init__.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+
+    global TEST_WORKSPACE
+    TEST_WORKSPACE = env.get_workspace('host_check')
```
Comment (gyorb) [inline tests/functional/host_check/__init__.py]: I do not think `TEST_WORKSPACE` setup is needed for these tests. `setup` and `teardown` could just `pass`.

File: tests/functional/ctu/ctu_failure/test_ctu_failure.py (hunk 12 above)
Comment (martong) [reply tests/functional/ctu/ctu_failure/test_ctu_failure.py]: Ok, I removed the cwd changing.

File: libcodechecker/analyze/host_check.py (hunk 11 above)
Comment (martong) [reply libcodechecker/analyze/host_check.py]: Ok, I added a `with` stmt.

File: tests/functional/host_check/__init__.py (hunk 14 above)
Comment (martong) [reply tests/functional/host_check/__init__.py]: Fixed.

File: tests/functional/host_check/__init__.py (hunk 13 above)
Comment (martong) [reply tests/functional/host_check/__init__.py]: Fixed.
